write more about BoundedCompactSupport and fun_prop (#584)
fpvandoorn/carleson: paper/main.tex

It is desirable for this simplicity to be reflected in the formalization.
This motivated packaging the recurring side conditions into test function classes.

> The structures \lean{BoundedCompactSupport} and \lean{BoundedFiniteSupport} contain essentially bounded measurable functions with compact support and with support of finite measure, respectively. A function $f$ is essentially bounded if there exists a real number $C$ such that $\|f(x)\|\le C$ almost everywhere.

Reviewer: "encode" essentially bounded measurable functions... maybe?

> The structures \lean{BoundedCompactSupport} and \lean{BoundedFiniteSupport} consist of functions that are essentially bounded measurable functions with compact support and with support of finite measure, respectively. A function $f$ is essentially bounded if there exists a real number $C$ such that $\|f(x)\|\le C$ almost everywhere.

Reviewer: nit: "of functions... that are ... functions" is repeating things; can you rephrase?

Lemmas such as \lean{hf.carlesonSum} show that the relevant
operators preserve the test function class, so once compact support has been established at the input, it remains available throughout the argument.

Choosing the right definition of these structures is subtle.
Since most functions we're working with are only well-behaved up to almost everywhere equality (see \ref{subsec:Lp}),
there is a choice in deciding that the functions are only essentially bounded or bounded everywhere.
Similarly, for \lean{BoundedCompactSupport} there is a choice in requiring that the function has compact support or essentially compact support

> (its support is a subset of the union of a compact set and a null set).

Reviewer: ```suggestion
(that is, its support is a subset of the union of a compact set and a null set).
```
Is this equivalent to "support is the union of a compact set and a null set"? (Or is the point that our measure need not be complete, i.e. not every subset of a null set must be even measurable?)

Author: It is equivalent to "it is the union of a relatively compact set (a set that is a subset of a compact set) and a null set". I'll reformualte it.

In Carleson, we originally started with functions that are bounded everywhere and have compact support,

> but we switched (in pull requests 311 and 342) to functions that are only essentially bounded,

Reviewer: re: "in pull requests 311 and 342"
I like recording this information --- but think we should present it differently. [We] [switched] with each word being a hyperlink could work, or adding these in a footnote.

Author: I'll move it to a footnote, and add links

but still have compact support.
The reason for this is that Mathlib uses essentially bounded functions frequently in measure theory,
phrased as \lean{MemLp f ∞ μ}.

These notions are mathematically very similar,
but the choice has impact in precisely which lemmas you can prove about it.
For example, if $g$ is bounded with compact support, and $f$ is a closed embedding,
is $g\circ f$ a bounded function with compact support?
This is only true if $g$ has compact support and is bounded everywhere;
otherwise $f$ may send a set of positive measure to a set of measure 0 where $g$ is ill-behaved.

> Another consideration into this design decision is using the functions in the class.

Reviewer: Can you tweak the wording "in the class"? The subsequent paragraph is very helpful, but this first sentence leaves more more puzzled than enlightened.

Author: is this better?

> Another consideration into this design decision is how nicely one can work with the functions in these structures.

Reviewer: Yes, better: I think "structures" reads strangely, though --- "classes" is better (though perhaps confusing in the context of a Lean paper). How about "how nicely one can work with these functions"? You're describing a mathematical issue next, not a Lean issue.

the argument consists of some tricky manipulations with suprema over the arguments,
and these would have to be changed to essential suprema, which are a bit trickier to work with.
In the end, we decided to prove that this function was both bounded with compact support,
and additionally bounded everywhere.

These classes are also tagged with the \lean{@[fun_prop]} attribute.
\lean{fun_prop} is a \mathlib-tactic developed by Tomáš Skřivan
that automatically uses tagged lemmas to show that a function has a certain property
(continuity, differentiability, measurability, etc.).
This tactic is good at writing complicated functions as a composition of simpler functions,
and then proving function properties by proving them first about the simpler functions.

\lean{fun_prop} is not as well-suited for proving integrability
as the other function properties mentioned above,
since the composition and product of integrable functions need not be integrable.
The class \lean{BoundedCompactSupport} is a step towards
making \texttt{fun\_prop} more usable for integrability,
since it is better behaved w.r.t. multiplication and composition.

> % maybe todo: mention the ICERM group on the integrability tactic?

Reviewer: Maybe? Let's discuss in person!

(that is, its support is the union of a relatively compact set and a null set,
where a relatively compact set is by definition a subset of a compact set).

In Carleson, we originally started with functions that are bounded everywhere and have compact support,
but we switched\footnote{see pull requests \href{https://github.com/fpvandoorn/carleson/pull/311}{311} and \href{https://github.com/fpvandoorn/carleson/pull/342}{342}} to functions that are only essentially bounded,
but still have compact support.
The reason for this is that Mathlib uses essentially bounded functions frequently in measure theory,
phrased as \lean{MemLp f ∞ μ}.

These notions are mathematically very similar,
but the choice has impact in precisely which lemmas you can prove about it.
For example, if $g$ is bounded with compact support, and $f$ is a closed embedding,
is $g\circ f$ a bounded function with compact support?
This is only true if $g$ has compact support and is bounded everywhere;
otherwise $f$ may send a set of positive measure to a set of measure 0 where $g$ is ill-behaved.

> Another consideration into this design decision is how nicely one can work with these function.

Reviewer: ```suggestion
Another consideration into this design decision is how nicely one can work with these functions.
```

> Another consideration into this design decision is how nicely one can work with the functions in these structures.

Author: ```suggestion
Another consideration into this design decision is how nicely one can work with these function.
```
